Fix cross-slice interaction detection for subplans

InnerLife0 opened the pull request:
GPDB has Shared Scan node to share the output of a subplan. Each Shared Scan node can act like producer or consumer. Each one may be a part of different slice. One of essential parts of Shared Scan processing is cross-slice interaction detection. It helps to find the difference in producer's and consumer's slices and mark underlying Material or Sort nodes as cross-slice. Cross-slice producer reads all the tuples from child plan before the consumer Shared Scan can read it.

Before this patch, cross-slice interaction detection of underlying subplan nodes was broken. We [walked](https://github.com/greenplum-db/gpdb/blob/6.20.5/src/backend/cdb/cdbmutate.c#L2642) through subplans separately from main plan. This, in case subplan has no motion, [caused](https://github.com/greenplum-db/gpdb/blob/6.20.5/src/backend/cdb/cdbmutate.c#L2488) `shareinput_mutator_xslice_2()` to think we don't need to mark node as cross-slice - motion stack (`motStack`) used for cross-slice interaction detection contained only [default](https://github.com/greenplum-db/gpdb/blob/6.20.5/src/backend/cdb/cdbmutate.c#L2628) motion id (0) which equals to producer's slice id.

Here is the simplified part of new regression test. It shows Shared Scan consumer under Subplan, which has different (compared to producer upper) slice id.
```
 Sequence
   ->  Shared Scan (share slice:id 0:0)
     ->  Seq Scan on t1 t1_1
   ->  Gather Motion 3:1  (slice3; segments: 3)
     ->  Redistribute Motion 1:3  (slice1)
       ->  Function Scan on generate_series
         SubPlan 1  (slice1)
           ->  Shared Scan (share slice:id 1:0)
```
Before this patch, we didn't have upper motions in `motStack`. From now, we dive into subplans as parts of main plan according to the plan tree hierarchy. Cross-slice detector understands that our slice(1) is different from producer's slice(0) and marks node as cross-slice.

Existing logic and comments around rtables left almost unchanged.
Existing tricky logic around joins execution order changed to another tricky logic which is compatible with `plan_tree_walker()`.
`gporca_optimizer.out` test output changed as we now start `shareinput_mutator_dag_to_tree()` from the main plan. Before, we started from [subplans](https://github.com/greenplum-db/gpdb/blob/6.20.5/src/backend/optimizer/plan/planner.c#L374) and global share input context contained only 3 producers at the time `set_plan_share_id()` [called](https://github.com/greenplum-db/gpdb/blob/6.20.5/src/backend/cdb/cdbmutate.c#L2145). Now, at the same place producers count is equal to 4, so RangeTblEntry's name changed.

Stolb27 commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -1772,135 +1772,133 @@ cdbmutate_warn_ctid_without_segid(struct PlannerInfo *root, struct RelOptInfo *r
-shareinput_walker(SHAREINPUT_MUTATOR f, Node *node, PlannerInfo *root)
+typedef struct ShareInputContext
+{
+	PlannerInfo *root;
```
First structure item for `plan_tree_walker` must be `plan_tree_base_prefix` (see this structure usage). Your code will work, but may create hidden bugs in case of refactoring plan_tree_walker.

InnerLife0 replied:
Yes, you are right.

Stolb27 commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -1772,135 +1772,133 @@ cdbmutate_warn_ctid_without_segid(struct PlannerInfo *root, struct RelOptInfo *r
+	SHAREINPUT_MUTATOR mutator;
+}	ShareInputContext;
+static bool
+shareinput_walker(Node *node, ShareInputContext * ctx)
```
Did you apply pgindent for your code? Uneven space formatting around *.

InnerLife0 replied:
Yes. And that change was made by pgindent.

Stolb27 commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -1772,135 +1772,133 @@ cdbmutate_warn_ctid_without_segid(struct PlannerInfo *root, struct RelOptInfo *r
+
+			/*
+			 * The code around rtables (for SubqueryScan and
+			 * TableFunctionScan) works only with appropriate subroot. Find
+			 * one and copy to 'ctx->root'. Use default, if not found.
+			 */
+			foreach(lc, root->glob->subroots)
+			{
+				PlannerInfo *subroot = (PlannerInfo *) lfirst(lc);
 
-			foreach(cell, mt->plans)
-				shareinput_walker(f, (Node *) lfirst(cell), root);
+				plan_id++;
+				if (plan_id == subplan->plan_id)
+				{
+					ctx->root = subroot;
+					break;
+				}
+			}
```
There is appropriate function `planner_subplan_get_root`. Am I correct?

InnerLife0 replied:
`planner_subplan_get_root` will fall to get out of range item.

Stolb27 replied:
Do you know cases when gpdb has a reference to not existing subplan?

InnerLife0 replied:
This from `shared_scan.sql`
```
drop table if exists pk_list;
create table pk_list (id int, schema_name varchar, table_name varchar) distributed by (id);
drop table if exists calender;
create table calender (id int, data_hour timestamp) distributed by (id);

explain (costs off)
with
	tbls as (select distinct schema_name, table_name as table_nm from pk_list),
	tbls_daily_report_23 as (select unnest(string_to_array('mart_cm.card' ,',')) as table_nm_23),
	tbls_w_onl_actl_data as (select unnest(string_to_array('mart_cm.cont_resp,mart_cm.card', ',')) as table_nm_onl_act)
select  data_hour, stat.schema_name as schema_nm, dt.table_nm
from (
	select * from calender c
	cross join tbls
) dt
inner join (
	select tbls.schema_name, tbls.table_nm as table_name
	from tbls tbls
) stat on dt.table_nm = stat.table_name
where
	(data_hour = date_trunc('day',data_hour) and stat.schema_name || '.' ||stat.table_name not in (select table_nm_23 from tbls_daily_report_23))
	and (stat.schema_name || '.' ||stat.table_name not in (select table_nm_onl_act from tbls_w_onl_actl_data))
	or (stat.schema_name || '.' ||stat.table_name in (select table_nm_onl_act from tbls_w_onl_actl_data));
```

InnerLife0 replied:
And this from `subselect_gp.sql`
```
CREATE TABLE foo_s (a integer, b integer)  PARTITION BY RANGE(b)
    (PARTITION sub_one START (1) END (10),
     PARTITION sub_two START (11) END (22));
INSERT INTO foo_s VALUES (9,9);
INSERT INTO foo_s VALUES (2,9);
CREATE TABLE bar_s (c integer, d character varying(10));
INSERT INTO bar_s VALUES (9,9);
CREATE TABLE baz_s (i int4);
INSERT INTO baz_s VALUES (9);

set optimizer=off;
SELECT bar_s.c FROM bar_s, foo_s WHERE foo_s.b = (SELECT max(i) FROM baz_s WHERE bar_s.c = 9) AND foo_s.b = bar_s.d::int4;
```

Stolb27 replied:
We found out that the only case when subplans and subroots arrays have inconsistent length is fixup_subplans modification. We are going to fix this problem too.

Stolb27 commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -1772,135 +1772,133 @@ cdbmutate_warn_ctid_without_segid(struct PlannerInfo *root, struct RelOptInfo *r
-				 * same as the subqueryscan's subplan, is valid
-				 * in Upstream but for not for GPDB, since we
-				 * create a new copy of the subplan if two
+				 * The Assert() on RelOptInfo's subplan being same as the
+				 * subqueryscan's subplan, is valid in Upstream but for not
+				 * for GPDB, since we create a new copy of the subplan if two
```
Maybe don't touch comments if we don't modify them to decrease patch footprint?

InnerLife0 replied:
I preferred to refactor all the function including it's comments. A lot of changes were done and it's painful to keep function partially formatted in such case.

Stolb27 commented on src/test/regress/sql/shared_scan.sql:
```diff
@@ -137,3 +137,78 @@ select case when ten < 5 then ten else ten * 2 end, count(distinct two), count(d
+select l.c from cte2 l
+left join t1 u on l.c = u.c;
+
+-- This case shows wrong count(*) result on pre-patched version.
+-- Shared Scan consumer from slice1 not marked as cross-slice and used not
+-- initialized tuple store. We got '1' as the result of the query - only
+-- Shared Scan from slice2 (the part below UNION) executed correctly.
+-- From now, cross-slice interaction detection fixed for subplans and we have
+-- stable '100' as a result.
```
This test case is flaky on unpatched version `PostgreSQL 9.4.26 (Greenplum Database 6.20.1_arenadata33 build dev)`:

```
--- \/home\/givi\/src\/gpdb6\/src\/test\/regress\/expected\/shared_scan_optimizer\.out  2022-06-17 10:00:40.097407850 +0000
+++ \/home\/givi\/src\/gpdb6\/src\/test\/regress\/results\/shared_scan\.out     2022-06-17 10:00:40.107407885 +0000
@@ -356,10 +356,7 @@
 )
 select l.c from cte2 l
 left join t1 u on l.c = u.c;
- c 
----
-(0 rows)
-
+ERROR:  cannot execute inactive Motion (nodeMotion.c:264) (nodeMotion.c:264)
 -- This case shows wrong count(*) result on pre-patched version.
 -- Shared Scan consumer from slice1 not marked as cross-slice and used not
 -- initialized tuple store. We got '1' as the result of the query - only

======================================================================
```

InnerLife0 replied:
It seems I was unclear. This test is flacky on old versions, because they miss new fault point. I'll change a comment.

Stolb27 replied:
My bad. It was a stupid question.

Commit:
Fix cross-slice interaction detection for subplans

GPDB has Shared Scan node to share the output of a subplan. Each Shared Scan
node can act like producer or consumer. Each one may be a part of different
slice. One of essential parts of Shared Scan processing is cross-slice
interaction detection. It helps to find the difference in producer's and
consumer's slices and mark underlying Material or Sort nodes as cross-slice.
Cross-slice producer reads all the tuples from child plan before the consumer
Shared Scan can read it.

Before this patch, cross-slice interaction detection of underlying subplan nodes
was broken. We walked through subplans separately from main plan. This, in case
subplan has no motion, caused `shareinput_mutator_xslice_2()` to think we don't
need to mark node as cross-slice - motion stack (`motStack`) used for
cross-slice interaction detection contained only default motion id (0) which
equals to producer's slice id.

Here is the simplified part of new regression test. It shows Shared Scan
consumer under Subplan, which has different (compared to producer upper) slice
id.
```
 Sequence
   ->  Shared Scan (share slice:id 0:0)
     ->  Seq Scan on t1 t1_1
   ->  Gather Motion 3:1  (slice3; segments: 3)
     ->  Redistribute Motion 1:3  (slice1)
       ->  Function Scan on generate_series
         SubPlan 1  (slice1)
           ->  Shared Scan (share slice:id 1:0)
```
Before this patch, we didn't have upper motions in `motStack`. From now, we dive
into subplans as parts of main plan according to the plan tree hierarchy.
Cross-slice detector understands that our slice(1) is different from producer's
slice(0) and marks node as cross-slice.

Existing logic and comments around rtables left almost unchanged.
Existing tricky logic around joins execution order changed to another tricky
logic which is compatible with `plan_tree_walker()`.
`gporca_optimizer.out` test output changed as we now start
`shareinput_mutator_dag_to_tree()` from the main plan. Before, we started from
subplans and global share input context contained only 3 producers at the time
`set_plan_share_id()` called. Now, at the same place producers count is equal to
4, so RangeTblEntry's name changed.

Commit:
Fix fixup_subplan_walker()

There was a bug when duplicate subplans were added without new corresponded
subroots. The old shareinput_walker() implementation not called for all
subplans, because forboth() loop stopped early. New shareinput_walker()
implementation not found appropriate subplan's subroot (and used main root as
default). From now, we fix old mistake and create new subroot for each new
subplan.
